Traktor Kontrol S3 script improvements, vanilla-like FX behavior, control initialization, better scratching, and more

robbert-vdh opened the pull request:
[Zullip thread](https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping/topic/Traktor.20Kontrol.20S3.20mapping.20improvements)

I've been using this for the last couple months and it's been a joy to use. This PR does depend on #11198 and #4651. The former exposes an option that should be enabled for the new vanilla FX behavior to work as intended, and latter is needed to be able to initialize the quick effect/filter knobs. Otherwise the effects are initialized after the control script has finished initializing and they'll be reset to their default values. This PR makes the following changes to the Traktor Kontrol S3 script:

- There's a new (enabled by default, but the old behavior is still available) effects implementation that closely matches the Traktor Mixer FX behavior that the controller was designed for (see [here](https://youtu.be/PU-sRgrlaJ8) for a demonstration). The idea is that each channel has one of five quick effect chain presets assigned. Clicking on the Filter button assigns the first quick effect chain preset, and clicking on the FX 1-4 buttons assigns the next four presets. The quick effect chain can then be controlled and toggled using the FX knob and FX On button on each channel. Holding one of the five FX/Filter select buttons and pressing a channel's FX On button will assign the effect to that channel without modifying the other channels. Everything is color coded and the FX buttons are highlighted just like in the video I linked above (which is for the S4 Mk3, but it works exactly the same here). For this to work best the super knob preservation option from #11198 should be enabled, the first quick effect preset chain should be Moog Filter or Filter, and the next four should be that same filter with an additional effect. I've made a couple of those, so if there's a good place to upload them (here?) I can do so and link them in the control script.
- When starting Mixxx/when reloading the script, Mixxx now queries the current knob positions for the mixer controls from the controller and initializes Mixxx's controls with them. For the filter/quick effect settings to restore correctly this way #4651 needs to be merged first.
- The default deck colors have been flipped to match the colors used in Traktor. Channels 1/2 are now blue, and channels 3/4 are orange by default.
- The FX On buttons in the new vanilla FX mode follow the quick effect chain preset assignments, but there's an off-by-default option to have them follow the track colors when the Filter/preset 1 has been assigned to the channel. This follows the current behavior, but in my experience using it it can get a bit confusing.
- Scratching is now implemented using the scratch2 methods and feels much smoother. Previously it would take about a second after releasing the jog wheel before playback started again.
- Added options (that don't do anything by default) for initializing deck parameters (sync, quantize, key lock, crossfader assignment). This is more part of my personal config so I can easily have the features I always use enabled when I start Mixxx so I can get going immediately. I can remove these options if desired.
- The halfway point for sliders/knobs is now correctly reported as 0.5. The S3 has an odd number of ticks for those controls, so setting a knob to the neutral position would actually send a value of `0.4998778998778999` to control object instead of 0.5. This would break certain features that explicitly check for 0.5 values, like the filters.
- The script has been modernized. Explicit prototypes have been replaced by ES6 classes, all uses of var except for the `TraktorS3` object have been removed (lots of variables previously used leaked variable bindings), and a custom bind function has been replaced with `fn.bind()`.

The manual still needs to be updated to mention the new FX behavior. Should I just create a PR in the manual repo?

robbert-vdh committed:
Use the engine scratch methods instead of setvalue

For the S3 controller mapping.

robbert-vdh committed:
Change default S3 deck colors to match Traktor

It's trivial to change this back if you prefer the calmer orange colors
for decks 1 and 2, but it seems like a good idea to stay as close to
stock as possible.

robbert-vdh committed:
Place S3 jog inertia behavior behind flag

I feel like by default it should just behave like everything else and
releasing the jog wheel should immediately resume playback.

robbert-vdh committed:
Completely remove jog inertia behavior from S3

I don't see why anyone would want to use this with scratch2 mode.

robbert-vdh committed:
Pass ticks to the scratch function for S3

Instead of the velocity from this manual velocity calculation.

robbert-vdh committed:
Use the correct channel number for scratching

Instead of the deck number.

robbert-vdh committed:
Remove paused jog multiplier in S3 script

Mixxx already does this internally.

robbert-vdh committed:
Allow initializing beatjump and beatloop sizes

robbert-vdh committed:
Allow initializing sync, quantize, keylock for S3

robbert-vdh committed:
Query current knob and slider values from the S3

This makes it so Mixxx's initial state matches the S3.

robbert-vdh committed:
Initialize packet parser with report data for S3

This makes sure no values can be lost.

robbert-vdh committed:
Add most of a Traktor-like alternative S3 FX impl

This behaves as intended by Native Instruments. Only the input parts
have been implemented, the LEDs still need to be done.

robbert-vdh committed:
Add an option to initialize S3 FX chains to filter

This makes more sense to me than keeping whatever was set last time.

robbert-vdh committed:
Add LED output for the new S3 FX behavior

This should now be pretty much done.

robbert-vdh committed:
Fix handling of multiple pressed FX select buttons

robbert-vdh committed:
Dimly light pressed FX Select buttons

And don't highlight the unpressed ones. This doesn't look as cool but it
makes it more obvious which ones are active.

robbert-vdh committed:
Rename StockFx to VanillaFx

robbert-vdh committed:
Rename lightFX() to lightFx()0

robbert-vdh committed:
Allow disabling the channel colors with vanilla FX

robbert-vdh committed:
Update the Traktor Kontrol S3 script authors

robbert-vdh committed:
Replace uses of TraktorS3.bind() with fn.bind()

robbert-vdh committed:
Convert S3 script to use ES6 classes

robbert-vdh committed:
Allow setting crossfader assignment from S3 script

robbert-vdh committed:
Replace S3 super knob hack with new control object

This is the same as `loaded_chain_preset`, except that it preserves the
value when changing presets. Bikeshedding over the name is very welcome.

robbert-vdh committed:
Replace var and global assignment in S3 script

With lexically scoped const/let bindings. The `var TraktorS3 = ...`
needs to stay a var by design.

robbert-vdh committed:
Fix 0.5 case in normalization of S3 values

This would make it impossible to set the quick effect super knob to
exactly 0.5. This is important for some effects like the Moog filter.

robbert-vdh committed:
Fix TraktorS3.VanillaFxModeChannelColors check

robbert-vdh committed:
Disable VanillaFxModeChannelColors by default

I liked the idea, but it quickly becomes confusing once you start
assigning effects to channels.

robbert-vdh committed:
Dimly light inactive FX select buttons

This is the original S3 behavior and it's probably less confusing.

robbert-vdh committed:
Adjust quick effect chain mappings in S3 script

As the result of https://github.com/mixxxdj/mixxx/pull/10859 being
merged. Everything shifted up one spot.

robbert-vdh committed:
Move effect chain preset offset to a constant

So this can be changed easier if this behavior ever changes again.

robbert-vdh committed:
Replace super knob preserving loaded_chain_preset

In the S3 script. There's now a preference that gives you the same
behavior.

robbert-vdh commented:
I'd also like to add controls for manual looping somewhere in the mapping. I've only needed this a couple times for a handful of tracks where fixed beatgrids just don't work (e.g. tempo changes, or the timing is just all over the place at different parts of the track). I searched a bit on the internet and I found people using  Shift+rev/flux for this, but when I gave that a try it felt very unergonomic because doing that with one hand while controlling the jog wheel with the other hand is a difficult exercise. Any objections to or better suggestions than using Shift+the deck switch buttons for this? That combination isn't used yet, and you can easily do this with one hand.

ywwg commented:
testing this finally! :)

ywwg commented:
With the new effect mode, something weird is happening.  I am trying to use just a regular Filter on the quick effect knob, not a filter echo.  If I select Filter from the list in the mixer UI, the effect light on the controller turns off and the filter knob works as expected. If I use the effect toggle on the controller again, it does toggle the effect in mixxx but the light doesn't change.  And if I push the effect buttons 1-4 or Filter, I can't get back to pure Filter mode.  It keeps selecting Filter+Echo.

We shouldn't try to match Traktor for deck colors.  Our skins use blue for decks C/D, so we should stick with that.  

Backspinning doesn't work that well any more, I don't get a fast spin out of it when I pull the wheel hard. The delay you mention was intentional to deal with cases where the dj has lifted their hand but the wheel is still spinning.  We should be directly driving the scratch2 speed parameter, [which directly corresponds to the track playback speed](https://github.com/mixxxdj/mixxx/blob/main/src/engine/controls/ratecontrol.cpp#L444-L447). The precise internal clock and high number of ticks means jog action should be extremely precise and shouldn't need smoothing, which causes lag.

ywwg reviewed (changes requested):
see above comment

ywwg commented:
(more on scratching -- the engine scratcher you used uses the computer clock, which is not the same as the controller's clock -- that's why it requires smoothing.  Because this controller has a sophisticated wheel reporting system, we don't need all of that.  That scratcher API is designed for cheaper controllers that only report ticks and no timing information)

JoergAtGithub commented:
> Backspinning doesn't work that well any more, I don't get a fast spin out of it when I pull the wheel hard. The delay you mention was intentional to deal with cases where the dj has lifted their hand but the wheel is still spinning.

Shouldn't this time depend on the parameters of engine.spinback:
  ```
/**
     * To achieve a spinback effect of the playback speed
     * Both engine.softStart() and engine.brake()/engine.spinback() can interrupt each other.
     * @param deck The deck number to use, e.g: 1
     * @param activate Set true to activate, or false to disable
     * @param factor Defines how quickly the deck should come to normal playback rate.
     *               Start with a value of 1 and increase to increase the acceleration.
     *               Be aware that spinback called with low factors (about 0.5 and lower),
     *               would keep the deck running altough the resulting very low sounds are not audible anymore. [default = 1.8]
     * @param rate The initial speed of the deck when enabled. "-10" (default) means 10x speed in reverse.
     *             Positive values like "10" also work, though then it's spinning forward obviously. [default = -10.0]
     */
    function spinback(deck:number, activate:boolean, factor?:number, rate?:number):void;
```
Maybe we need an optional callback function for `engine.spinback`, which is called when normal playback rate is reached after spinning?

robbert-vdh committed:
Revert "Change default S3 deck colors to match Traktor"

This reverts commit e38537e99a3e57c5d02424213037883097ebfc5c.

robbert-vdh commented:
> With the new effect mode, something weird is happening. I am trying to use just a regular Filter on the quick effect knob, not a filter echo. If I select Filter from the list in the mixer UI, the effect light on the controller turns off and the filter knob works as expected. If I use the effect toggle on the controller again, it does toggle the effect in mixxx but the light doesn't change. And if I push the effect buttons 1-4 or Filter, I can't get back to pure Filter mode. It keeps selecting Filter+Echo.

You'll need to go to Settings -> Effects and move the Filter effect to the top of the quick effect chain presets list.

> We shouldn't try to match Traktor for deck colors. Our skins use blue for decks C/D, so we should stick with that.

Okay, I'll revert the change.

> Backspinning doesn't work that well any more, I don't get a fast spin out of it when I pull the wheel hard.

Controller jog wheels don't usually spin as much as CDJ jog wheels. Those use tracks with actual wheels, which is why they can spin for so long if you turn the tension all the way down. So with these kinds of jog wheels you kind of need to give it some, uh, assistance and manually spin the wheel back a bit if you want a truly impressive pullup. If you stop playback before releasing the platter the behavior will be the same as before, but I really don't think playback should stop completely for a while before abruptly starting again when you release the platter. Have you tried manually beatmatching with the controller? Or just navigating around tracks? For anything other than hard pullups having it behave more like vinyl and having playback immediately ramp up again to the target rate feels very natural and seems like the correct behavior to me. That's also what the Traktor S2 scripts do (they have the exact same jogwheel as the S3). And while the jog wheel may have 768 ticks (which in the grand scheme of thing isn't terribly precise, the DDJ-400 uses 720 ticks), almost every discrete signal can benefit from being low-passed to avoid discontinuities (for this to work even better Mixxx's audio engine should also be chopping the buffer up into smaller 64 sample blocks). You can control the filter's parameters using the alpha and beta constants in the script.

Also, there is one remaining gremlin I've run into twice now but that I can't easily reproduce (but I haven't really tried either). When the volume faders are set to the lowest value when you start Mixxx, sometimes the soft takeover doesn't work correctly. The fader will be initialized all the way at the bottom, but moving it doesn't do anything. You need to move it down and then up again for it to work.

ronso0 commented:
> Our skins use blue for decks C/D

Hmm? LateNight C/D is blue-ish, besides that I'm not aware of any color rules.
Anyway, controller deck colors could be made customizable + a hint in the mapping's `<description>` element.

ronso0 commented:
> > Backspinning doesn't work that well any more, I don't get a fast spin out of it when I pull the wheel hard. The delay you mention was intentional to deal with cases where the dj has lifted their hand but the wheel is still spinning.
> 
> Shouldn't this time depend on the parameters of engine.spinback

`engine.spinback` is just an emulation that can be triggered with buttons.
@ywwg is referring to the 'real' way ; )

robbert-vdh commented:
> > Our skins use blue for decks C/D
> 
> Hmm? LateNight C/D is blue-ish, besides that I'm not aware of any color rules. Anyway, controller deck colors could be made customizable + a hint in the mapping's `<description>` element.

If that is the case (I have no idea) I'd prefer sticking with the original blue/orange (for primary/secondary) color scheme. It's easy enough to change, and that makes it just a bit more recognizable for any Traktor refugees considering switching to Mixxx.

ywwg commented:
> Hmm? LateNight C/D is blue-ish, besides that I'm not aware of any color rules.

Deere also uses blue for C/D, so it seemed to be a consistent choice.  If our two most popular skins use blue for C/D it seems like the controller should as well

ywwg commented:
> You can control the filter's parameters using the alpha and beta constants in the script.

but the fact remains that if you use the engine.enableScratch API like you did, you're throwing out the [controller's timer](https://github.com/mixxxdj/mixxx/blob/main/res/controllers/Traktor-Kontrol-S3-hid-scripts.js#L700) and using the [less-precise mixxx timer](https://github.com/mixxxdj/mixxx/blob/main/src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp#L681), and this will cause much more imprecision in the speed detection. If you want to use smoothing that's fine, but you'll have to implement that in javascript yourself.  Please revert to the behavior that drives scratch2 directly. If you want to change the code that [starts a timer when the user lifts their hand off the wheel](https://github.com/mixxxdj/mixxx/blob/main/res/controllers/Traktor-Kontrol-S3-hid-scripts.js#L653-L658), that's fine.

JoergAtGithub commented:
Wouldn't it make sense to add an optional argument for an external time reference to scratch2?
HID controllers might be a special case here, but MIDI should provide time stamps for jog position data.

robbert-vdh commented:
> Deere also uses blue for C/D, so it seemed to be a consistent choice. If our two most popular skins use blue for C/D it seems like the controller should as well

Mixxx supports more controllers with switchable decks, do they all use orange/blue for primary/secondary?

> but the fact remains that if you use the engine.enableScratch API like you did, you're throwing out the [controller's timer](https://github.com/mixxxdj/mixxx/blob/main/res/controllers/Traktor-Kontrol-S3-hid-scripts.js#L700) and using the [less-precise mixxx timer](https://github.com/mixxxdj/mixxx/blob/main/src/controllers/scripting/legacy/controllerscriptinterfacelegacy.cpp#L681), and this will cause much more imprecision in the speed detection. If you want to use smoothing that's fine, but you'll have to implement that in javascript yourself. Please revert to the behavior that drives scratch2 directly. If you want to change the code that [starts a timer when the user lifts their hand off the wheel](https://github.com/mixxxdj/mixxx/blob/main/res/controllers/Traktor-Kontrol-S3-hid-scripts.js#L653-L658), that's fine.

I think there are some misunderstandings with how the jog wheel currently works and how the controller scripts interact with the audio engine. The old behavior (where `scratch2` is controlled manually from the script) is inherently broken, and while `engine.scratchTick()` may not be perfect, it mostly gets it right.

- A core fundamental problem with driving the `scratch2` control object directly from a controller script is that controller scripts don't (and since they're event based, can't) run in lockstep with the audio thread. The audio thread is the heart of an audio application, and it's the one place where precision matters. The audio thread's process function is called once per audio buffer, or period (either term is used depending on the platform). During that process function Mixxx computes the playback rate (and like I previously mentioned, this should be done multiple times for smaller blocks) using `RateControl::calculateSpeed()`. That function in turn uses the value stored in the `scratch2` control object to directly control the playback rate if scratching is enabled. Where I'm going with this is that the `scratch2` value is only accurate if it is computed immediately before or during `RateControl::calculateSpeed()`. It thus must be computed on the audio thread, or old values are (re)used.
- That aforementioned process function is called once per period/buffer of audio that must be produced. That means that the interval between process calls depends on both the period size and the sample rate. For smaller period sizes like 128 samples it will be called more frequently than for larger period sizes like 8192. The `TraktorS3.Deck.prototype.jogHandler()` function from the (older) S3 control script meanwhile is called whenever a HID message that changes the jog wheel's position is received. This rate is not tied to the audio thread's process callbacks. When rotating the jog wheel at a certain pace it may be called twice in between process calls, or there may be two process calls between `jogHandler()` calls. Setting the `scratch2` value from `jogHandler()` thus always results in incorrect behavior. Here's an experiment:
  - Set the buffer size to 128 samples, or the lowest latency setting. On the current 2.4 branch the JACK period size is not used, so use ALSA on Linux, ASIO on Windows, or CoreAudio on macOS. Select a track, jump to a cue point, put your finger on the 6 o' clock position on the jogwheel, or wherever you prefer, and then as slowly as possible make one turn clockwise. Then make one full counter clockwise turn, as fast as you can. Note that the playback position is now slightly _after_ the original cue point. If you keep repeating these slow clockwise and fast counterclockwise turns you'll see the playhead drift more and more to the right.
  - Now do the same thing with the buffer size set to 8192 samples, or the highest latency setting. You'll notice that the drifting has become even worse. This is because the HID thread still runs at the same pace, while the audio thread's pace has slowed down. More of the `scratch2` messages during the fast counterclockwise turns are being ignored.
  - Finally hold down the platter, slowly rotate it, and then hold still without releasing the platter. You'll see that the playhead continues to travel in the direction you spun the jog wheel in. This is because the `scratch2` value is set in the control script, and it's then never reset to 0.
- The only way to get this behavior right, is to _accumulate_ the scratch ticks (which is what `engine.scratchTick()` does), and to use those accumulated ticks as an input for a smoothing filter _on the audio thread_ (which is also what the `engine.scratch*()` family of functions do).
- The timing of the ticks as reported by the controller doesn't matter much in the end. What matters is that if n ticks are received, n ticks should also have been handled by the end of the smoothing. That's a property of most (or more or less all) minimum-phase low-pass filters's step responses. If you did somehow incorporate the timing and have ticks be processed earlier, you'd get discontinuities which the low-pass filter specifically aims to reduce.

So I strongly believe the new behavior from this PR is the correct one.

ywwg commented:
If the amount of jog wheel drift is less with your version than mine, then I have to concede it's the better approach.  I am wondering why NI chose to incorporate a high-precision timer in their controller if it does not help improve accuracy?  Perhaps they poll the value inside the audio callback?  That is out of scope for this change, I am just trying to figure it out.  I worked hard to reverse engineer the HID spec so I am sad to toss it out :(.  I would also prefer to increase the sensitivity so that spinbacks sound better.  I am not a scratch DJ so I don't use jog for much else, and I don't want to use the javascript spinback.  Maybe sensitivity can be a user parameter at the top of the file?

I want to the color scheme as-is for this PR because it seems that we need to think about unifying the color choices across controllers and skins. I see the argument for maintaining consistency with Traktor, but with three of our skins using blue for the 3/4 decks, it does not make sense to use blue for decks A/B. We should not have the default be so unintuitive. The current state:
latenight classic: decks 1/2: neutral, decks 3/4: blue
latenight palemoon: decks 1/2: neutral, decks 3/4: blue
deere: decks 1/2: neutral, decks 3/4: blue
Shade: does not support 4 decks
Tango: decks 1/2: purple, decks 3/4: green 

Furthermore it is not common for users to migrate from a paid product like Traktor over to Mixxx. More commonly a user is already on Mixxx or Virtual DJ and will buy a used S3, and not seeing Traktor at all. Mixxx is in the unfortunate position of needing to play nice with many different DJ platforms, so as a policy we don't use any one software as a reference.  So let's leave this alone for now and come back to it in a subsequent PR.

ywwg reviewed (changes requested):
Can we also add a sensitivity constant for the jog wheels?

robbert-vdh committed:
Add a jog wheel multiplier constant for the S3

robbert-vdh commented:
> Can we also add a sensitivity constant for the jog wheels?

Sure! I thought about it, but when I checked all other scripts just stick with the mapping to 33 1/3 rpm vinyl so I went with that. I've added a constant to the script for configuring this.

While I did this I noticed another issue that affects every control script. They all set the `jog` control object value for a channel when nudging. This value is supposed to be an accumulator that's reset to 0 after being read. The current approach has both a precision issue and a data race issue. If the buffer size is larger than the value may be overwritten multiple times before it is read. This can be fixed by using an atomic fetch-and-add operation instead of overwriting the value. The second issue is more widespread in Mixxx and that is the fact that the value is read, compared, and then set to 0 if it was nonzero. Between that read and set a new value could have been written to it by the control script. This can be fixed by using an atomic exchange instead. I'll create an issue for this if I don't forget.

robbert-vdh commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -15,6 +17,47 @@
+// The first mode is dubbed 'vanilla mode' as it behaves in the way the mixer FX
+// section was originally intended to be used by Native Instruments. Disable
+// this option to use the second, Mixxx-specific mode.
+TraktorS3.VanillaFxMode = true;
```
Basic implies limited and advanced implies being a superset of the basic functionality with more options exposed. That probably gives the wrong impression. What about effect chain mode (or some wording containing quick effects)?

ywwg replied:
So do you mean "BasicFxMode" and "EffectChainMode"?  

other ideas: "SingleFxMode" vs "MultiFxMode"; "QuickEffectsMode" vs "EffectChainMode"  or really any permutation of those haha.

robbert-vdh replied:
I'd probably go for quick effects mode and multi effect mode. Effect chain also doesn't seem like the right choice because the new behavior lets you toggle between effect chains while the old behavior uses fixed effect chains.

ywwg reviewed (approved):
not blocking on the name bikeshedding, LGTM :)

robbert-vdh commented:
Yeah naming things is one of the most difficult things in software engineering for a reason. ;) I think I like effect chain mode for the new behavior and multi FX mode or indeed just advanced mode for the old behavior. I also still like vanilla mode (in previous commits I referred to it as stock FX mode) since it's close to the behavior someone would expect after either having seen videos of it on YouTube or after having used it with Traktor would expect. I'll update the manual once we settle on names heh.

ywwg commented:
As a follow-up to this can you also update the manual entry for the S3 to describe the functional changes you made?

ronso0 commented:
A ready-to-be-merged manual PR adding / updating the  mapping documentation is mandatory for mapping changes to be merged btw.

robbert-vdh commented:
> A ready-to-be-merged manual PR adding / updating the mapping documentation is mandatory for mapping changes to be merged btw.

I was still waiting for any suggested behavior and name changes before making the manual changes so I didn't have to do it twice heh. I'll get to it now.

ronso0 commented:
Sure, no hurry, don't worry : )
That was just ment as a reminder that we set this rule to avoid confusion & complaints in case the mapping changed  and the docu doesn't reflect that, yet.

robbert-vdh committed:
Change 'Vanilla Mode' to 'Quick Effect Mode'

robbert-vdh commented:
I created https://github.com/mixxxdj/manual/pull/533 for the manual changes.

ywwg commented:
As noted in the other thread, this change is blocked on https://github.com/mixxxdj/mixxx/pull/11198

ywwg commented:
It would be nice to get as many of these fixes in as we can, would it be possible to split out the changes that require #11198 so we can get this in now, and then commit the other changes later?

robbert-vdh commented:
On a side note, I noticed the other day that some of the LED state isn't synchronized when changing between layers. I noticed that the play/pause button blinking didn't match with the new channel just after switching, and there may be other controls that also don't match. I have not yet had the time to look into this.

ywwg commented:
are those regressions, or is that the same as the current state?  As long as the experience on the controller is strictly better with this PR, I'd like to get it in and keep iterating

robbert-vdh commented:
I haven't tested the current version but I also didn't touch that part of the code (other than transforming the prototypes to ES6 classes), so it shouldn't be a regression.

robbert-vdh commented:
Yeah they're on my personal todo list. I'll get to them (and perhaps open PRs) after I finish a couple other things.

JoergAtGithub commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -60,2149 +96,2160 @@ TraktorS3.ChannelColors = {
+        // with different values once. Report 2 contains the state of the mixer
+        // controls.
+        const report2Values = new Uint8Array(controller.getInputReport(2));
+        TraktorS3.incomingData([2, ...Array.from(report2Values.map(x => ~x))]);
+        TraktorS3.incomingData([2, ...Array.from(report2Values)]);
```
```suggestion
        TraktorS3.incomingData(new Uint8Array([2, ...Uint8Array.from(report2Values.map(x => ~x))]));
        TraktorS3.incomingData(new Uint8Array([2, ...Uint8Array.from(report2Values)]));
```
When incomingData is called from C++, the argument is an array of type Uint8Array. The same type should be used here too.

robbert-vdh replied:
Done.

robbert-vdh committed:
Pass Uint8Arrays to TraktorS3.incomingData

As suggested in
https://github.com/mixxxdj/mixxx/pull/11199#discussion_r1133314678. The
function does accept both typed arrays and plain arrays.

robbert-vdh commented:
Oh yeah, there's one other thing I was wondering about with this mapping. This line from the previous version of the script creates a link from the `[ChannelN],scratch2_enable` control object to change the a deck's lighting. Mixxx warns that this control object doesn't actually exist. Was this added in error or was the control object renamed at some point? I think I did a quick git pickaxe search at one point but that didn't turn up anything particularly interesting.

JoergAtGithub commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -60,2149 +96,2160 @@ TraktorS3.ChannelColors = {
+        // just tell the packet parser the current values so it won't ignore the
+        // next input.
+        const report1Values = new Uint8Array(controller.getInputReport(1));
+        TraktorS3.incomingData([1, ...Array.from(report1Values)]);
```
```suggestion
        TraktorS3.incomingData(new Uint8Array([1, ...Uint8Array.from(report1Values)]));
```
This should also be Uint8Array

robbert-vdh replied:
Fixed. The HID parser previously accepted both regular arrays and Uint8Array but after #4894 it only accepts Uint8Arrays.

ywwg commented:
thanks for working on the docs!

JoergAtGithub commented:
As far as I can see the changes to the FX behavior block merging this PR since a while. Would it make sense to move th FX changes to a follow-up PR and merge the other improvements into 2.4?

robbert-vdh commented:
Without the option from #11198 this would still work. It could just lead to some slightly confusing behavior if you've ever used Pioneer, Denon, NI, or other DJ gear with channel FX selectors before.

JoergAtGithub commented:
@ywwg What is needed to get this merged?

JoergAtGithub commented:
Merge this without the super-knob default PR, because this is independent: https://github.com/mixxxdj/mixxx/pull/11198#issuecomment-1633720880
The text section in the manual PR got adjusted therefore.
Thank you!
